doc/batcher.8
.\" $Revision$
.TH BATCHER 8
.SH NAME
batcher \- article batching backend for InterNetNews
.SH SYNOPSIS
.B batcher
[
.BI \-a " arts"
]
[
.BI \-A " total_arts"
]
[
.BI \-b " size"
]
[
.BI \-B " total_size"
]
[
.BI \-i " string"
]
[
.BI \-N " num_batches"
]
[
.BI \-p " process"
]
[
.B \-r
]
[
.BI \-s " separator"
]
[
.BI \-S " alt_spool"
]
[
.B \-v
]
.I host
[
.I input
]
.SH DESCRIPTION
.I Batcher
reads uses a list of files to prepare news batches for the specified
.IR host .
It is normally invoked by a script run out of
.IR cron (8)
that uses
.IR shlock (1)
to lock the host name, followed by a
.IR ctlinnd (8)
command to flush the batchfile.
.PP
.I Batcher
reads the named
.I input
file, or standard input if no file is given.
Relative pathnames are interpreted from the
.I <pathoutgoing in inn.conf>
directory.
The input is taken as a set of lines.
Blank lines and lines starting with a number sign (``#'') are ignored.
All other lines should consist of one or two fields separated by a single space.
The first field is the name of a file holding an article; if it is not an
an absolute pathname it is taken relative to the news spool directory,
.IR <patharticles\ in\ inn.conf> .
The second field, if present, specifies the size of the article in bytes.
.SH OPTIONS
.TP
.B \-S
The ``\-S'' flag may be used to specify an alternate spool directory to
use if the article is not found; this would normally be an NFS-mounted
spool directory of a master server with longer expiration times.
.TP
.B \-r
By default, the program sets its standard error to
.IR <pathlog\ in\ inn.conf>/errlog .
To suppress this redirection, use the ``\-r'' flag.
.TP
.B \-v
Upon exit,
.I batcher
reports statistics via
.IR syslog (3).
If the ``\-v'' flag is used, they will also be printed on the standard
output.
.TP
.B \-b
.I Batcher
collects the text of the named articles into batches.
To limit the size of each batch, use the ``\-b'' flag.
The default size is 60 kilobytes.
Using ``\-b0'' allows unlimited batch sizes.
.TP
.B \-a
To limit the number of articles in each batch, use the ``\-a'' flag.
The default is no limit.
A new batch will be started when either the byte count or number of
articles written exceeds the specified limits.
.TP
.B \-B
To limit the total number of bytes written for all batches, use the ``\-B''
flag.
.TP
.B \-A
To limit the total number of articles that can be batched use the ``\-A''
flag.
.TP
.B \-N
To limit the total number of batches that should be created use the ``\-N''
flag.
.IP
In all three cases, the default is zero, which is taken to mean no limit.
.TP
.B \-i string
A batch starts with an identifying line to specify the unpacking method
to be used on the receiving end.
When the ``\-i'' flag is used, the initial string,
.IR string ,
followed by a newline, will be output at the start of every batch.
The default is to have no initial string.
.TP
.B \-s
Each article starts with a separator line to indicate the size of the article.
To specify the separator use the ``\-s'' flag.
This is a
.IR sprintf (3)
format string which can have a single ``%ld'' parameter which will be given
the size of the article.
If the separator is not empty, then the string and a newline will be output
before every article.
The default separator is ``#! rnews %ld''.
.TP
.B \-p
By default, batches are written to standard output, which
is not useful when more than one output batch is created.
Use the ``\-p'' flag to specify the shell command that should be
created (via
.IR popen (3))
whenever a new batch is started.
The process is a
.I sprintf
format string which can have a single ``%s'' parameter which will be given
the host name.
A common value is:
.PP
.RS
.nf
( echo '#! cunbatch' ; exec compress ) | uux \- \-r \-z %s!rnews
.fi
.RE
.SH EXIT STATUS
.PP
If the input is exhausted,
.I batcher
will exit with a zero status.
If any of the limits specified with the ``\-B,'' ``\-A,'' or ``\-N'' flags
is reached, or if there is an error writing the batch, then
.I batcher
will try to spool the input, copying it to a file.
If there was no input filename, the standard input will be copied to
.I <pathoutgoing in inn.conf>/host
and the program will exit.
If an input filename was given, a temporary file named
.IR input .bch
(if
.I input
is an absolute pathname)
or
.I <pathoutgoing in inn.conf>/input.bch
(if the filename does not begin with a slash) is created.
Once the input is copied,
.I batcher
will try to rename this temporary file to be the name of the input file,
and then exit.
.PP
Upon receipt of an interrupt or termination signal,
.I batcher
will finish sending the current article, close the batch, and then
rewrite the batchfile according as described in the previous paragraph.
.SH HISTORY
Written by Rich $alz <rsalz@uunet.uu.net> for InterNetNews.
.de R$
This is revision \\$3, dated \\$4.
..
.R$ $Id$
.SH "SEE ALSO"
inn.conf(5),
ctlinnd(8),
newsfeeds(5),
shlock(1).


NEWS
Current

    Here are the changes for the latest version of INN.

    * New readers.conf file (replaces nnrp.access) which allows more
      flexible specification of access restrictions.

    * Unified overview has been replaced with an overview API, and there are
      now two separate overview implementations to choose from.  One
      (tradindexed) is very like traditional overview but uses an additional
      index file for faster responses to readers.  The other (buffindexed)
      uses circular buffers and can handle a higher incoming article rate
      while still being fast for readers.  The ovmethod key in inn.conf
      chooses the overview method to use.

    * All article storage and retrieval is now done via the storage API. 
      Traditional spool is now available as a storage type under the storage
      API.

    * The timecaf storage method has been added, similar to timehash but
      storing multiple articles in a single file.  See INSTALL for details
      on it.

    * INN now supports embedded Python filters as well as Perl and TCL
      filters.

Version 2.2

    * New storage.conf file (replaces storage.ctl).

    * New (optional) way of handling non-cancel control messages
      (controlchan) that serializes them and prevents server overload from
      control message storms.

    * Support for actsyncd to fetch active file with ftp; configured by
      default to get the one at ftp://ftp.isc.org/pub/usenet/CONFIG/active.Z
      if you run actsyncd.  Be sure to read the manual page for actsync to
      configure an actsync.ign file for your site, and test simpleftp if you
      do not configure with wget or ncftp.  Also see
      ftp://ftp.isc.org/pub/usenet/CONFIG/README.

    * Some options to configure are now moved to inn.conf (merge-to-groups
      and pgp-verify).

    * inndf, a portable version of df(1) is supplied.

    * New cnfsstat program to show stats of cnfs buffers.

    * news2mail program for gateway'ing news into email is provided.

    * mailpost program for gateway'ing email to news is provided.

    * pullnews program for doing a sucking feed is provided (not meant for
      large feeds).

    * The innshellvars.csh.in script is obsolete (and lives in the obsolete
      directory, for now).



backends/batcher.c
/*  $Id$
**
**  Read batchfiles on standard input and spew out batches.
*/

#include "config.h"
#include "clibrary.h"
#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <syslog.h> 
#include <sys/stat.h>

#include "inn/innconf.h"
#include "inn/messages.h"
#include "inn/timer.h"
#include "inn/wire.h"
#include "inn/libinn.h"
#include "inn/paths.h"
#include "inn/storage.h"


/*
**  Global variables.
*/
static bool	BATCHopen;
static double	STATbegin;
static double	STATend;
static char	*Host;
static char	*InitialString;
static char	*Input;
static char	*Processor;
static int	ArtsInBatch;
static size_t	ArtsWritten;
static int	BATCHcount;
static int	MaxBatches;
static int	BATCHstatus;
static size_t	BytesInBatch = 60 * 1024;
static size_t	BytesWritten;
static size_t	MaxArts;
static size_t	MaxBytes;
static sig_atomic_t	GotInterrupt;
static const char *Separator = "#! rnews %ld";
static char	*ERRLOG;

/*
**  Start a batch process.
*/
static FILE *
BATCHstart(void)
{
    FILE	*F;
    char	buff[SMBUF];

    if (Processor && *Processor) {
#pragma GCC diagnostic ignored "-Wformat-nonliteral"
	snprintf(buff, sizeof(buff), Processor, Host);
#pragma GCC diagnostic warning "-Wformat-nonliteral"
	F = popen(buff, "w");
	if (F == NULL)
	    return NULL;
    }
    else
	F = stdout;
    BATCHopen = true;
    BATCHcount++;
    return F;
}


/*
**  Close a batch, return exit status.
*/
static int
BATCHclose(FILE *F)
{
    BATCHopen = false;
    if (F == stdout)
	return fflush(stdout) == EOF ? 1 : 0;
    return pclose(F);
}


/*
**  Update the batch file and exit.
*/
static void
RequeueAndExit(off_t Cookie, char *line, long BytesInArt)
{
    char	*spool;
    char	buff[BIG_BUFFER];
    int		i;
    FILE	*F;
    double	usertime;
    double	systime;

    /* Do statistics. */
    STATend = TMRnow_double();
    if (GetResourceUsage(&usertime, &systime) < 0) {
	usertime = 0;
	systime = 0;
    }

    notice("batcher %s times user %.3f system %.3f elapsed %.3f",
           Host, usertime, systime, STATend - STATbegin);
    notice("batcher %s stats batches %d articles %lu bytes %lu",
           Host, BATCHcount, (unsigned long) ArtsWritten,
           (unsigned long) BytesWritten);

    /* Last batch exit okay? */
    if (BATCHstatus == 0) {
	if (feof(stdin) && Cookie != -1) {
	    /* Yes, and we're all done -- remove input and exit. */
	    fclose(stdin);
	    if (Input)
		unlink(Input);
	    exit(0);
	}
    }

    /* Make an appropriate spool file. */
    if (Input == NULL)
        spool = concatpath(innconf->pathoutgoing, Host);
    else
        spool = concat(Input, ".bch", (char *) 0);
    if ((F = xfopena(spool)) == NULL)
        sysdie("%s cannot open %s", Host, spool);

    /* If we can back up to where the batch started, do so. */
    i = 0;
    if (Cookie != -1 && fseeko(stdin, Cookie, SEEK_SET) == -1) {
        syswarn("%s cannot seek", Host);
	i = 1;
    }

    /* Write the line we had; if the fseeko worked, this will be an
     * extra line, but that's okay. */
    if (line && fprintf(F, "%s %ld\n", line, BytesInArt) == EOF) {
        syswarn("%s cannot write spool", Host);
	i = 1;
    }

    /* Write rest of stdin to spool. */
    while (fgets(buff, sizeof buff, stdin) != NULL) 
	if (fputs(buff, F) == EOF) {
            syswarn("%s cannot write spool", Host);
	    i = 1;
	    break;
	}
    if (fclose(F) == EOF) {
        syswarn("%s cannot close spool", Host);
	i = 1;
    }

    /* If we had a named input file, try to rename the spool. */
    if (Input != NULL && rename(spool, Input) < 0) {
        syswarn("%s cannot rename spool", Host);
	i = 1;
    }

    free(spool);
    exit(i);
    /* NOTREACHED */
}


/*
**  Mark that we got interrupted.
*/
static void
CATCHinterrupt(int s)
{
    GotInterrupt = true;

    /* Let two interrupts kill us. */
    xsignal(s, SIG_DFL);
}


int
main(int ac, char *av[])
{
    bool	Redirect;
    FILE	*F;
    char	*p;
    char	line[BIG_BUFFER];
    char	buff[BIG_BUFFER];
    size_t	BytesInArt;
    size_t	BytesInCB;
    off_t	Cookie;
    int		i;
    int		ArtsInCB;
    int		length;
    TOKEN	token;
    ARTHANDLE	*art;
    char	*artdata;
    size_t      written;

    /* Set defaults. */
    openlog("batcher", L_OPENLOG_FLAGS | LOG_PID, LOG_INN_PROG);
    message_program_name = "batcher";
    if (!innconf_read(NULL))
        exit(1);
    Redirect = true;
    umask(NEWSUMASK);
    ERRLOG = concatpath(innconf->pathlog, INN_PATH_ERRLOG);

    /* By default, statistics only go to syslog. */
    message_handlers_notice(1, message_log_syslog_notice);

    /* Parse JCL. */
    while ((i = getopt(ac, av, "a:A:b:B:i:N:p:rs:v")) != EOF)
	switch (i) {
	default:
            die("usage error");
            break;
	case 'a':
	    ArtsInBatch = atoi(optarg);
	    break;
	case 'A':
	    MaxArts = atol(optarg);
	    break;
	case 'b':
	    BytesInBatch = atol(optarg);
	    break;
	case 'B':
	    MaxBytes = atol(optarg);
	    break;
	case 'i':
	    InitialString = optarg;
	    break;
	case 'N':
	    MaxBatches = atoi(optarg);
	    break;
	case 'p':
	    Processor = optarg;
	    break;
	case 'r':
	    Redirect = false;
	    break;
	case 's':
	    Separator = optarg;
	    break;
	case 'v':
            message_handlers_notice(2, message_log_syslog_notice,
                                    message_log_stdout);
	    break;
	}
    if (MaxArts && ArtsInBatch == 0)
	ArtsInBatch = MaxArts;
    if (MaxBytes && BytesInBatch == 0)
	BytesInBatch = MaxBytes;

    /* Parse arguments. */
    ac -= optind;
    av += optind;
    if (ac != 1 && ac != 2)
        die("usage error");
    Host = av[0];
    if ((Input = av[1]) != NULL) {
	if (Input[0] != '/')
            Input = concatpath(innconf->pathoutgoing, av[1]);
	if (freopen(Input, "r", stdin) == NULL)
            sysdie("%s cannot open %s", Host, Input);
    }

    if (Redirect)
	freopen(ERRLOG, "a", stderr);

    /* Set initial counters, etc. */
    BytesInCB = 0;
    ArtsInCB = 0;
    Cookie = -1;
    GotInterrupt = false;
    xsignal(SIGHUP, CATCHinterrupt);
    xsignal(SIGINT, CATCHinterrupt);
    xsignal(SIGTERM, CATCHinterrupt);
    /* xsignal(SIGPIPE, CATCHinterrupt); */
    STATbegin = TMRnow_double();

    SMinit();
    F = NULL;
    while (fgets(line, sizeof line, stdin) != NULL) {
	/* Record line length in case we do an ftello.  Not portable to
	 * systems with non-Unix file formats. */
	length = strlen(line);
	Cookie = ftello(stdin) - length;

	/* Get lines like "name size".  Note that we ignore size but accept
         * it for backwards compatibility. */
	if ((p = strchr(line, '\n')) == NULL) {
            warn("%s skipping %.40s: too long", Host, line);
	    continue;
	}
	*p = '\0';
	if (line[0] == '\0' || line[0] == '#')
	    continue;
        p = strchr(line, ' ');
        if (p != NULL)
            *p = '\0';

        p = line;

	/* Open the article. */
	if (IsToken(p)) {
	    token = TextToToken(p);
	    if ((art = SMretrieve(token, RETR_ALL)) == NULL) {
		if ((SMerrno != SMERR_NOENT) && (SMerrno != SMERR_UNINIT))
                    warn("%s skipping %.40s: %s", Host, p, SMerrorstr);
		continue;
	    }
	    artdata = wire_to_native(art->data, art->len, &BytesInArt);
	    SMfreearticle(art);
	} else {
            warn("%s skipping %.40s: not token", Host, p);
	    continue;
	}

	/* Have an open article, do we need to open a batch?  This code
	 * is here (rather than up before the while loop) so that we
	 * can avoid sending an empty batch.  The goto makes the code
	 * a bit more clear. */
	if (F == NULL) {
	    if (GotInterrupt) {
		RequeueAndExit(Cookie, (char *)NULL, 0L);
	    }
	    if ((F = BATCHstart()) == NULL) {
                syswarn("%s cannot start batch %d", Host, BATCHcount);
		break;
	    }
	    if (InitialString && *InitialString) {
		fprintf(F, "%s\n", InitialString);
		BytesInCB += strlen(InitialString) + 1;
		BytesWritten += strlen(InitialString) + 1;
	    }
	    goto SendIt;
	}

	/* We're writing a batch, see if adding the current article
	 * would exceed the limits. */
	if ((ArtsInBatch > 0 && ArtsInCB + 1 >= ArtsInBatch)
	 || (BytesInBatch > 0 && BytesInCB + BytesInArt >= BytesInBatch)) {
	    if ((BATCHstatus = BATCHclose(F)) != 0) {
		if (BATCHstatus == -1)
                    syswarn("%s cannot close batch %d", Host, BATCHcount);
		else
                    syswarn("%s batch %d exit status %d", Host, BATCHcount,
                            BATCHstatus);
		break;
	    }
	    ArtsInCB = 0;
	    BytesInCB = 0;

	    /* See if we can start a new batch. */
	    if ((MaxBatches > 0 && BATCHcount >= MaxBatches)
	     || (MaxBytes > 0 && BytesWritten + BytesInArt >= MaxBytes)
	     || (MaxArts > 0 && ArtsWritten + 1 >= MaxArts)) {
		break;
	    }

	    if (GotInterrupt) {
		RequeueAndExit(Cookie, (char *)NULL, 0L);
	    }

	    if ((F = BATCHstart()) == NULL) {
                syswarn("%s cannot start batch %d", Host, BATCHcount);
		break;
	    }
	}

    SendIt:
	/* Now we can start to send the article! */
	if (Separator && *Separator) {
#pragma GCC diagnostic ignored "-Wformat-nonliteral"
	    snprintf(buff, sizeof(buff), Separator, BytesInArt);
#pragma GCC diagnostic warning "-Wformat-nonliteral"
	    BytesInCB += strlen(buff) + 1;
	    BytesWritten += strlen(buff) + 1;
	    if (fprintf(F, "%s\n", buff) == EOF || ferror(F)) {
                syswarn("%s cannot write separator", Host);
		break;
	    }
	}

        /* Write the article.  In case of interrupts, retry the read but not
         * the fwrite because we can't check that reliably and portably. */
        written = fwrite(artdata, 1, BytesInArt, F);
	if (written != BytesInArt || ferror(F))
	    break;

	/* Update the counts. */
	BytesInCB += BytesInArt;
	BytesWritten += BytesInArt;
	ArtsInCB++;
	ArtsWritten++;

	if (GotInterrupt) {
	    Cookie = -1;
	    BATCHstatus = BATCHclose(F);
	    RequeueAndExit(Cookie, line, BytesInArt);
	}
    }

    if (BATCHopen)
	BATCHstatus = BATCHclose(F);
    RequeueAndExit(Cookie, NULL, 0);

    return 0;
}
